packages/assets-controllers/src/NftController.ts

```diff
@@ -667,46 +857,49 @@ export class NftController extends BaseControllerV1<NftConfig, NftState> {
           nftMetadata,
           existingEntry,
         );
-        if (differentMetadata || !existingEntry.isCurrentlyOwned) {
-          // TODO: Switch to indexToUpdate
-          const indexToRemove = nfts.findIndex(
-            (nft) =>
-              nft.address.toLowerCase() === tokenAddress.toLowerCase() &&
-              nft.tokenId === tokenId,
-          );
-          /* istanbul ignore next */
-          if (indexToRemove !== -1) {
-            nfts.splice(indexToRemove, 1);
-          }
-        } else {
-          return nfts;
+
+        if (!differentMetadata && existingEntry.isCurrentlyOwned) {
+          return;
         }
-      }
 
-      const newEntry: Nft = {
-        address: tokenAddress,
-        tokenId,
-        favorite: existingEntry?.favorite || false,
-        isCurrentlyOwned: true,
-        ...nftMetadata,
-      };
+        const indexToUpdate = nfts.findIndex(
+          (nft) =>
+            nft.address.toLowerCase() === checksumHexAddress.toLowerCase() &&
+            nft.tokenId === tokenId,
+        );
+
+        if (indexToUpdate !== -1) {
+          nfts[indexToUpdate] = {
+            ...existingEntry,
+            ...nftMetadata,
+          };
+        }
+      } else {
+        const newEntry: Nft = {
+          address: checksumHexAddress,
+          tokenId,
+          favorite: false,
+          isCurrentlyOwned: true,
+          ...nftMetadata,
+        };
+
+        nfts.push(newEntry);
+      }
 
-      const newNfts = [...nfts, newEntry];
-      this.updateNestedNftState(newNfts, ALL_NFTS_STATE_KEY, {
+      this.#updateNestedNftState(nfts, 'allNfts', {
```

Comment:
nit: For consistency 

```suggestion
      this.#updateNestedNftState(nfts, ALL_NFTS_STATE_KEY, {
```